### DISCUSSION TITLE
How to use package that depends on redux-saga when its added to webpack-node-externals allowList

### DISCUSSION THREAD

**Discussion by krutoo** [webpack/webpack#17743] [Q&A] [answered]
> Hi, i ask same question in redux-saga repo but i think there is some build step details
> https://github.com/redux-saga/redux-saga/issues/2396

I'm having difficulty building a project for Node.js via webpack.

In my case in the project I use my own package that uses redux-saga as dependency

In webpack I specified that my package should be included in the bundle via webpack-node-externals allowList option.

After build and run app in Node.js it crashes with error exactly in the place where my package uses redux-saga, at this place error is: createSagaMiddleware is not a function

In the same time if i import redux-saga directly in project just like it's done in package - it will works perfectly

I see that CJS and ESM versions of redux-saga is different because if i just copy code that uses redux-saga from my package to project - it will works fine

### Repro

I prepare repo with exmaple of issue:
https://github.com/krutoo/webpack-externals-breaks-isomorph

just install dependencies and run npx webpack watch --mode=development

### Details

As i understand:

- external code uses `require('redux-saga')` that use `exports` object as function instead its "default" property
- bundled code uses ESM version that uses default import as function

**Comment by krutoo** [comment]
its not helps when i use `import * as ReduxSaga from 'redux-saga'; ReduxSaga.default(...);`

**Reply by alexander-akait** [reply, marked as answer]
You have `import createSagaMiddleware from "redux-saga";` but after compilation you have `require("redux-saga")` and `redux-sage` use `default` for constructor, but there is no default in commonjs (in commonjs you have an export object just with the `default` property), that is why you have such problems and that is why named export are recommended for libraries.  

Anyway we have a solution for it:
```js
nodeExternals({
      importType: function (moduleName) {
        if (/^redux-saga$/.test(moduleName)) {
          return ['commonjs ' + moduleName, 'default'];
        }

        return 'commonjs ' + moduleName;
      },
      // this next line breaks build
      allowlist: [/(@sima-land)/],
    }),
```

Sorry for delay and feel free to feedback
